Foundation/UserDefaults.swift

```diff
@@ -469,30 +349,44 @@ open class UserDefaults: NSObject {
     
     open func setPersistentDomain(_ domain: [String : Any], forName domainName: String) {
         if let defaults = UserDefaults(suiteName: domainName) {
+            var removedAny = false
             for key in defaults._dictionaryRepresentation(searchingOutsideOfSuite: false).keys {
                 defaults.removeObject(forKey: key)
+                removedAny = true
             }
             
+            var addedAny = false
             for (key, value) in domain {
                 defaults.set(value, forKey: key)
+                addedAny = true
+            }
+            
+            _ = defaults.synchronize()
+            
+            if removedAny || addedAny {
+                NotificationCenter.default.post(name: UserDefaults.didChangeNotification, object: self)
             }
         }
-        
-        NotificationCenter.default.post(name: UserDefaults.didChangeNotification, object: self)
     }
     
     open func removePersistentDomain(forName domainName: String) {
         if let defaults = UserDefaults(suiteName: domainName) {
+            var removedAny = false
             for key in defaults._dictionaryRepresentation(searchingOutsideOfSuite: false).keys {
                 defaults.removeObject(forKey: key)
+                removedAny = true
             }
             
-            NotificationCenter.default.post(name: UserDefaults.didChangeNotification, object: self)
+            _ = defaults.synchronize()
+            
+            if removedAny {
```

Comment:
This notification is posted unconditionally in the ObjC implementation

Reply:
Will change.